codec-doh/src/main/java/io/netty/handler/codec/doh/DohRecordEncoder.java

```diff
@@ -0,0 +1,117 @@
+import io.netty.handler.codec.http.DefaultFullHttpRequest;
+import io.netty.handler.codec.http.HttpHeaderNames;
+import io.netty.handler.codec.http.HttpMethod;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpVersion;
+import io.netty.handler.codec.http.QueryStringEncoder;
+import io.netty.util.internal.ObjectUtil;
+
+import java.net.InetSocketAddress;
+import java.util.Base64;
+
+public final class DohRecordEncoder extends ChannelOutboundHandlerAdapter {
+    private static final String DEFAULT_DOH_PATH = "/dns-query";
+    private final DohQueryEncoder dohQueryEncoder = new DohQueryEncoder();
+
+    private final InetSocketAddress dohServer;
+    private final boolean useHttpPost;
+    private final String uri;
+
+    /**
+     * Creates a new instance.
+     *
+     * @param dohServer the dohServer address
+     */
+    public DohRecordEncoder(InetSocketAddress dohServer) {
+        this(dohServer, true, DEFAULT_DOH_PATH);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param dohServer the dohServer address
+     * @param useHttpPost the http request method that can be used to connect to dohServer
+     */
+    public DohRecordEncoder(InetSocketAddress dohServer, boolean useHttpPost) {
+        this(dohServer, useHttpPost, DEFAULT_DOH_PATH);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param dohServer the dohServer address
+     * @param uri the http request uri that can be used as address path
+     */
+    public DohRecordEncoder(InetSocketAddress dohServer, String uri) {
+        this(dohServer, true, uri);
+    }
+
+    /**
+     * Creates a new instance.
+     *
+     * @param dohServer the dohServer address
+     * @param useHttpPost the http request method that can be used to connect to dohServer
+     * @param uri the http request uri that can be used as address path
+     */
+    public DohRecordEncoder(InetSocketAddress dohServer, boolean useHttpPost, String uri) {
+        this.dohServer = ObjectUtil.checkNotNull(dohServer, "dohServer");
+        this.useHttpPost = useHttpPost;
+        this.uri = ObjectUtil.checkNotNull(uri, "uri");
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+        ByteBuf content = ctx.alloc().buffer();
+        dohQueryEncoder.encode(ctx, (DnsQuery) msg, content);
+
+        HttpRequest request = useHttpPost ? createPostRequest(content, uri) : createGetRequest(content, uri);
+
+        request.headers().set(HttpHeaderNames.HOST, dohServer.getHostName());
+        request.headers().set(HttpHeaderNames.ACCEPT, "application/dns-message");
+        request.headers().set(HttpHeaderNames.CONTENT_TYPE, "application/dns-message");
+
+        if (useHttpPost) {
+            request.headers().set(HttpHeaderNames.CONTENT_LENGTH, content.readableBytes());
+        }
+
+        super.write(ctx, request, promise);
+    }
```

Comment:
We also need to release the query and the content buffer, otherwise we will leak.

```suggestion
    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
        DnsQuery query = (DnsQuery) msg
        ByteBuf content = ctx.alloc().buffer();
        try {
            dohQueryEncoder.encode(ctx, query, content);

            HttpRequest request = useHttpPost ? createPostRequest(content, uri) : createGetRequest(content, uri);

            request.headers().set(HttpHeaderNames.HOST, dohServer.getHostName());
            request.headers().set(HttpHeaderNames.ACCEPT, "application/dns-message");
            request.headers().set(HttpHeaderNames.CONTENT_TYPE, "application/dns-message");

            if (useHttpPost) {
                request.headers().set(HttpHeaderNames.CONTENT_LENGTH, content.readableBytes());
           }

           ctx.write(request, promise);
        } finally {
            content.release();
            query.release();
        }
    }
```